client/clientutil/modelinfo.go

```diff
@@ -0,0 +1,580 @@
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package clientutil
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"strings"
+	"text/tabwriter"
+	"time"
+
+	"github.com/snapcore/snapd/asserts"
+	"github.com/snapcore/snapd/i18n"
+	"github.com/snapcore/snapd/strutil"
+	"github.com/snapcore/snapd/timeutil"
+)
+
+var (
+	invalidTypeMessage = i18n.G("invalid type for %q header")
+	errNoMainAssertion = errors.New(i18n.G("device not ready yet (no assertions found)"))
+	errNoSerial        = errors.New(i18n.G("device not registered yet (no serial assertion found)"))
+
+	// this list is a "nice" "human" "readable" "ordering" of headers to print
+	// off, sorted in lexical order with meta headers and primary key headers
+	// removed, and big nasty keys such as device-key-sha3-384 and
+	// device-key at the bottom
+	// it also contains both serial and model assertion headers, but we
+	// follow the same code path for both assertion types and some of the
+	// headers are shared between the two, so it still works out correctly
+	niceOrdering = [...]string{
+		"architecture",
+		"base",
+		"classic",
+		"display-name",
+		"gadget",
+		"kernel",
+		"revision",
+		"store",
+		"system-user-authority",
+		"timestamp",
+		"required-snaps", // for uc16 and uc18 models
+		"snaps",          // for uc20 models
+		"device-key-sha3-384",
+		"device-key",
+	}
+)
+
+type ModelFormatter interface {
+	GetPublisher() string
+	GetEscapedDash() string
+}
+
+// modelAssertJSON is a helper to write out a single assertion in json
+// format that matches how it's done in api_model.go. We have stolen
+// the structure from there.
+type modelAssertJSON struct {
+	Headers map[string]interface{} `json:"headers,omitempty"`
+	Body    string                 `json:"body,omitempty"`
+}
+
+func fmtTime(t time.Time, abs bool) string {
+	if abs {
+		return t.Format(time.RFC3339)
+	}
+	return timeutil.Human(t)
+}
+
+// PrintModelAssertionYAML will format the provided serial or model assertion based on the parameters given in
+// YAML format. The output will be written to the provided io.Writer.
+func PrintModelAssertionYAML(w *tabwriter.Writer, modelFormatter ModelFormatter, termWidth int, useSerial, absTime, verbose, assertion bool, modelAssertion *asserts.Model, serialAssertion *asserts.Serial) error {
+	var mainAssertion asserts.Assertion
+
+	// if we got an invalid model assertion bail early
+	if modelAssertion == nil {
+		return errNoMainAssertion
+	}
+
+	if useSerial {
```

Comment:
once this code moves to a different package useSerial is awkward, this relates also to my comment about errors, I would have two functions:

PrimtModelAssertionYAML that takes a model and serial,  the model is mandatory and can't be nil
PrintSerialAssertionYAML  takes a serial and the serial cannot be nil

Reply:
PrintSerialAssertionYAML also needs the Model assertion, so they would take same parameters, except we remove the useSerial bool? That doesn't seem like a good tradeoff?

Reply:
I'm quite sure that the serial can be printed without the model, if we use the model when printing the serial is probably just a code quirk that can be fixed. If you find that's not the case I would be curious to know.

Reply:
Grade and storageSafety comes from modelAssertation
```
	grade := modelAssertion.HeaderString("grade")
	storageSafety := modelAssertion.HeaderString("storage-safety")
```